File: Back-End/common/validators.py
"""
Common reusable validators for DWMS.

All validators raise `rest_framework.serializers.ValidationError` so they can be
used directly inside DRF serializer `validate_*` / `validate()` methods.

Each validator returns the cleaned / normalised value on success — apply on the
write path with `value = validate_xxx(value)`.

Naming convention:
    validate_<thing>(value, *, allow_blank=False) -> str | Decimal | ...

Indian-fintech focused; safe defaults; conservative regexes (rejects clearly
invalid inputs, not "creative-but-plausible" ones).
"""

from __future__ import annotations

import re
from decimal import Decimal
from typing import Optional

from rest_framework import serializers


# ──────────────────────────────────────────────────────────────────────────────
# Regex constants (compiled once)
# ──────────────────────────────────────────────────────────────────────────────

_RE_USERNAME      = re.compile(r'^[A-Za-z][A-Za-z0-9._-]{2,49}$')
_RE_PASSWORD_LOW  = re.compile(r'[a-z]')
_RE_PASSWORD_UP   = re.compile(r'[A-Z]')
_RE_PASSWORD_DIG  = re.compile(r'\d')
_RE_PASSWORD_SYM  = re.compile(r'[^A-Za-z0-9]')

_RE_INDIAN_MOBILE = re.compile(r'^[6-9]\d{9}$')             # 10-digit, starts 6-9
_RE_PAN           = re.compile(r'^[A-Z]{5}\d{4}[A-Z]$')
_RE_IFSC          = re.compile(r'^[A-Z]{4}0[A-Z0-9]{6}$')   # 11 chars, 5th is 0
_RE_GSTIN         = re.compile(r'^\d{2}[A-Z]{5}\d{4}[A-Z]\d[A-Z][A-Z\d]$')
_RE_UPI_VPA       = re.compile(r'^[a-zA-Z0-9._-]{2,50}@[a-zA-Z][a-zA-Z0-9.-]{1,40}$')
_RE_ACCOUNT_NO    = re.compile(r'^\d{9,18}$')
_RE_SAFE_NAME     = re.compile(r"^[A-Za-z0-9 .,'&()\-_/]+$")
_RE_BRAND_NAME    = re.compile(r'^[A-Z0-9 _-]{2,50}$')
_RE_ROLE_NAME     = re.compile(r'^[A-Za-z][A-Za-z0-9 _-]{1,49}$')
_RE_CLIENT_ARC    = re.compile(r'^[A-Za-z0-9_-]{3,50}$')


# ──────────────────────────────────────────────────────────────────────────────
# Generic helpers
# ──────────────────────────────────────────────────────────────────────────────

def _clean(value, *, allow_blank: bool = False, field: str = 'value') -> str:
    if value is None:
        if allow_blank:
            return ''
        raise serializers.ValidationError(f'{field} is required.')
    s = str(value).strip()
    if not s and not allow_blank:
        raise serializers.ValidationError(f'{field} cannot be blank.')
    return s


# ──────────────────────────────────────────────────────────────────────────────
# Identity / contact validators
# ──────────────────────────────────────────────────────────────────────────────

def validate_username(value: str) -> str:
    """3–50 chars, letters/digits/._- only, must start with a letter."""
    s = _clean(value, field='Username')
    if not _RE_USERNAME.match(s):
        raise serializers.ValidationError(
            'Username must be 3–50 characters, start with a letter, and contain only letters, digits, dot, underscore or hyphen.'
        )
    return s


def validate_indian_mobile(value: Optional[str], *, allow_blank: bool = True) -> str:
    """10-digit Indian mobile starting with 6-9. Accepts +91 / 91 / 0 prefixes; returns the 10-digit form."""
    if value in (None, ''):
        if allow_blank:
            return ''
        raise serializers.ValidationError('Mobile number is required.')
    digits = re.sub(r'\D', '', str(value))
    # strip leading country code / trunk prefix
    if len(digits) == 12 and digits.startswith('91'):
        digits = digits[2:]
    elif len(digits) == 11 and digits.startswith('0'):
        digits = digits[1:]
    if not _RE_INDIAN_MOBILE.match(digits):
        raise serializers.ValidationError(
            'Enter a valid 10-digit Indian mobile number (starting with 6-9).'
        )
    return digits


def validate_password_strength(value: str, *, min_length: int = 8) -> str:
    """Strong password: ≥ min_length, has upper, lower, digit, and symbol."""
    if value is None or len(value) < min_length:
        raise serializers.ValidationError(
            f'Password must be at least {min_length} characters long.'
        )
    checks = [
        (_RE_PASSWORD_LOW, 'at least one lowercase letter'),
        (_RE_PASSWORD_UP,  'at least one uppercase letter'),
        (_RE_PASSWORD_DIG, 'at least one digit'),
        (_RE_PASSWORD_SYM, 'at least one symbol'),
    ]
    missing = [label for rx, label in checks if not rx.search(value)]
    if missing:
        raise serializers.ValidationError(
            'Password must include ' + ', '.join(missing) + '.'
        )
    if ' ' in value:
        raise serializers.ValidationError('Password must not contain spaces.')
    return value


# ──────────────────────────────────────────────────────────────────────────────
# Government / banking identifiers (India)
# ──────────────────────────────────────────────────────────────────────────────

def validate_pan(value: str) -> str:
    """PAN: 5 letters + 4 digits + 1 letter (uppercase). Returns upper-cased."""
    s = _clean(value, field='PAN').upper()
    if not _RE_PAN.match(s):
        raise serializers.ValidationError('Enter a valid PAN (e.g. ABCDE1234F).')
    return s


def validate_ifsc(value: str) -> str:
    """IFSC: 11 chars — 4 letters + 0 + 6 alphanumeric. Returns upper-cased."""
    s = _clean(value, field='IFSC code').upper().replace(' ', '')
    if not _RE_IFSC.match(s):
        raise serializers.ValidationError(
            'Enter a valid 11-character IFSC code (e.g. HDFC0001234).'
        )
    return s


def validate_gstin(value: str) -> str:
    s = _clean(value, field='GSTIN').upper().replace(' ', '')
    if not _RE_GSTIN.match(s):
        raise serializers.ValidationError('Enter a valid 15-character GSTIN.')
    return s


def validate_upi_vpa(value: str) -> str:
    """UPI VPA / handle, e.g. name@bank."""
    s = _clean(value, field='UPI ID').lower().replace(' ', '')
    if not _RE_UPI_VPA.match(s):
        raise serializers.ValidationError(
            'Enter a valid UPI ID (e.g. name@bank).'
        )
    return s


def validate_account_number(value: str) -> str:
    """Bank account number — digits only, 9–18 characters."""
    s = re.sub(r'\D', '', _clean(value, field='Account number'))
    if not _RE_ACCOUNT_NO.match(s):
        raise serializers.ValidationError(
            'Enter a valid bank account number (9 to 18 digits).'
        )
    return s


# ──────────────────────────────────────────────────────────────────────────────
# Domain name validators (brands / roles / clients)
# ──────────────────────────────────────────────────────────────────────────────

def validate_brand_name(value: str) -> str:
    """Brand code/name: uppercase letters/digits/space/_/-, 2–50 chars."""
    s = _clean(value, field='Brand name').upper().strip()
    if not _RE_BRAND_NAME.match(s):
        raise serializers.ValidationError(
            'Brand name must be 2–50 uppercase characters (letters, digits, space, underscore or hyphen).'
        )
    return s


def validate_role_name(value: str) -> str:
    s = _clean(value, field='Role name').strip()
    if not _RE_ROLE_NAME.match(s):
        raise serializers.ValidationError(
            'Role name must be 2–50 characters, start with a letter, and contain only letters, digits, space, underscore or hyphen.'
        )
    return s


def validate_safe_name(value: str, *, field: str = 'Name', max_length: int = 150) -> str:
    """Generic safe name — letters/digits and a few common punctuation."""
    s = _clean(value, field=field)
    if len(s) > max_length:
        raise serializers.ValidationError(f'{field} must be at most {max_length} characters.')
    if not _RE_SAFE_NAME.match(s):
        raise serializers.ValidationError(
            f'{field} contains invalid characters.'
        )
    return s


def validate_client_arc_id(value: str) -> str:
    s = _clean(value, field='Client ARC ID')
    if not _RE_CLIENT_ARC.match(s):
        raise serializers.ValidationError(
            'Client ARC ID must be 3–50 chars: letters, digits, underscore or hyphen.'
        )
    return s


# ──────────────────────────────────────────────────────────────────────────────
# Numeric / amount validators
# ──────────────────────────────────────────────────────────────────────────────

def validate_positive_amount(value, *, field: str = 'Amount', max_value: Optional[Decimal] = Decimal('10000000.00')) -> Decimal:
    """Positive decimal amount, ≤ max_value (default 1 crore)."""
    if value is None:
        raise serializers.ValidationError(f'{field} is required.')
    try:
        amt = Decimal(str(value))
    except Exception:
        raise serializers.ValidationError(f'{field} must be a valid number.')
    if amt <= 0:
        raise serializers.ValidationError(f'{field} must be greater than zero.')
    if max_value is not None and amt > max_value:
        raise serializers.ValidationError(f'{field} cannot exceed {max_value:,.2f}.')
    if amt.as_tuple().exponent < -2:
        raise serializers.ValidationError(f'{field} can have at most 2 decimal places.')
    return amt


def validate_range(range_from, range_to, *, field_to: str = 'range_to') -> None:
    """Ensure range_from < range_to (both Decimal-ish)."""
    if range_from is None or range_to is None:
        return
    try:
        a, b = Decimal(str(range_from)), Decimal(str(range_to))
    except Exception:
        raise serializers.ValidationError({field_to: 'Range values must be valid numbers.'})
    if a < 0 or b < 0:
        raise serializers.ValidationError({field_to: 'Range values cannot be negative.'})
    if a >= b:
        raise serializers.ValidationError({field_to: f'{field_to} must be greater than range_from.'})


# ──────────────────────────────────────────────────────────────────────────────
# Text / message validators
# ──────────────────────────────────────────────────────────────────────────────

def validate_text(value, *, field: str = 'Text', max_length: int = 1000, allow_blank: bool = True, min_length: int = 0) -> str:
    """Trim, length-check generic free-text input."""
    s = _clean(value, allow_blank=allow_blank, field=field)
    if allow_blank and not s:
        return s
    if len(s) < min_length:
        raise serializers.ValidationError(f'{field} must be at least {min_length} characters.')
    if len(s) > max_length:
        raise serializers.ValidationError(f'{field} must be at most {max_length} characters.')
    return s

File: Front-End/src/utils/validators.js
// ─────────────────────────────────────────────────────────────────────────────
// Shared field validators — mirrors Back-End/common/validators.py rules.
// Each validator returns null on success or a human-readable error string.
// ─────────────────────────────────────────────────────────────────────────────

// ── Regex constants (compiled once) ─────────────────────────────────────────
export const RX = {
  username:      /^[A-Za-z][A-Za-z0-9._-]{2,49}$/,
  indianMobile:  /^[6-9]\d{9}$/,
  pan:           /^[A-Z]{5}\d{4}[A-Z]$/,
  ifsc:          /^[A-Z]{4}0[A-Z0-9]{6}$/,
  gstin:         /^\d{2}[A-Z]{5}\d{4}[A-Z]\d[A-Z][A-Z\d]$/,
  upi:           /^[a-zA-Z0-9._-]{2,50}@[a-zA-Z][a-zA-Z0-9.-]{1,40}$/,
  accountNumber: /^\d{9,18}$/,
  brandName:     /^[A-Z0-9 _-]{2,50}$/,
  roleName:      /^[A-Za-z][A-Za-z0-9 _-]{1,49}$/,
  safeName:      /^[A-Za-z0-9 .,'&()\-_/]+$/,
  clientArcId:   /^[A-Za-z0-9_-]{3,50}$/,
}

// ── File size & mime ────────────────────────────────────────────────────────
export const FILE = {
  IMAGE_EXTS:  ['.jpg', '.jpeg', '.png', '.webp', '.gif', '.bmp', '.tiff', '.tif'],
  PDF_EXTS:    ['.pdf'],
  DOC_EXTS:    ['.doc', '.docx', '.xls', '.xlsx', '.csv', '.txt'],
  LIMITS_MB: {
    image: 5, pdf: 8, slip: 8, attachment: 15, qr: 5,
  },
}

const _ext = (name) => {
  const i = (name || '').lastIndexOf('.')
  return i === -1 ? '' : name.slice(i).toLowerCase()
}

// ── Generic ─────────────────────────────────────────────────────────────────
export const required = (label = 'This field') => (v) =>
  v === null || v === undefined || String(v).trim() === '' ? `${label} is required.` : null

export const maxLen = (n, label = 'Value') => (v) =>
  v && String(v).length > n ? `${label} must be at most ${n} characters.` : null

export const minLen = (n, label = 'Value') => (v) =>
  v && String(v).length < n ? `${label} must be at least ${n} characters.` : null

// ── Identity / contact ──────────────────────────────────────────────────────
export const username = (v) => {
  if (!v) return 'Username is required.'
  return RX.username.test(v)
    ? null
    : 'Username must be 3–50 characters, start with a letter, and contain only letters, digits, dot, underscore or hyphen.'
}

export const indianMobile = (v, { allowBlank = true } = {}) => {
  if (!v) return allowBlank ? null : 'Mobile number is required.'
  let digits = String(v).replace(/\D/g, '')
  if (digits.length === 12 && digits.startsWith('91')) digits = digits.slice(2)
  else if (digits.length === 11 && digits.startsWith('0')) digits = digits.slice(1)
  return RX.indianMobile.test(digits)
    ? null
    : 'Enter a valid 10-digit Indian mobile number (starting with 6-9).'
}

export const strongPassword = (v, { minLength = 8 } = {}) => {
  if (!v || v.length < minLength) return `Password must be at least ${minLength} characters long.`
  const missing = []
  if (!/[a-z]/.test(v)) missing.push('a lowercase letter')
  if (!/[A-Z]/.test(v)) missing.push('an uppercase letter')
  if (!/\d/.test(v))    missing.push('a digit')
  if (!/[^A-Za-z0-9]/.test(v)) missing.push('a symbol')
  if (missing.length) return `Password must include ${missing.join(', ')}.`
  if (/\s/.test(v)) return 'Password must not contain spaces.'
  return null
}

// ── Banking ─────────────────────────────────────────────────────────────────
export const pan = (v) => {
  if (!v) return 'PAN is required.'
  return RX.pan.test(String(v).toUpperCase())
    ? null
    : 'Enter a valid PAN (e.g. ABCDE1234F).'
}

export const ifsc = (v) => {
  if (!v) return 'IFSC code is required.'
  return RX.ifsc.test(String(v).toUpperCase().replace(/\s/g, ''))
    ? null
    : 'Enter a valid 11-character IFSC code (e.g. HDFC0001234).'
}

export const upi = (v) => {
  if (!v) return 'UPI ID is required.'
  return RX.upi.test(String(v).toLowerCase().replace(/\s/g, ''))
    ? null
    : 'Enter a valid UPI ID (e.g. name@bank).'
}

export const accountNumber = (v) => {
  if (!v) return 'Account number is required.'
  const digits = String(v).replace(/\D/g, '')
  return RX.accountNumber.test(digits)
    ? null
    : 'Enter a valid bank account number (9 to 18 digits).'
}

export const gstin = (v) => {
  if (!v) return 'GSTIN is required.'
  return RX.gstin.test(String(v).toUpperCase().replace(/\s/g, ''))
    ? null
    : 'Enter a valid 15-character GSTIN.'
}

// ── Domain names ────────────────────────────────────────────────────────────
export const brandName = (v) => {
  if (!v) return 'Brand name is required.'
  return RX.brandName.test(String(v).toUpperCase().trim())
    ? null
    : 'Brand name must be 2–50 uppercase characters (letters, digits, space, underscore or hyphen).'
}

export const roleName = (v) => {
  if (!v) return 'Role name is required.'
  return RX.roleName.test(v)
    ? null
    : 'Role name must be 2–50 characters, start with a letter, and contain only letters, digits, space, underscore or hyphen.'
}

export const safeName = (v, label = 'Name', max = 150) => {
  if (!v) return `${label} is required.`
  if (v.length > max) return `${label} must be at most ${max} characters.`
  return RX.safeName.test(v) ? null : `${label} contains invalid characters.`
}

export const clientArcId = (v) => {
  if (!v) return 'Client ARC ID is required.'
  return RX.clientArcId.test(v)
    ? null
    : 'Client ARC ID must be 3–50 chars: letters, digits, underscore or hyphen.'
}

// ── Numeric ─────────────────────────────────────────────────────────────────
export const positiveAmount = (v, { label = 'Amount', max = 10_000_000 } = {}) => {
  if (v === null || v === undefined || v === '') return `${label} is required.`
  const n = Number(v)
  if (Number.isNaN(n)) return `${label} must be a valid number.`
  if (n <= 0) return `${label} must be greater than zero.`
  if (max != null && n > max) return `${label} cannot exceed ${max.toLocaleString('en-IN')}.`
  // 2 decimal places max
  const s = String(v)
  if (s.includes('.') && s.split('.')[1].length > 2) {
    return `${label} can have at most 2 decimal places.`
  }
  return null
}

export const rangeOrder = (from, to, label = 'Max amount') => {
  if (from == null || to == null || from === '' || to === '') return null
  return Number(from) >= Number(to) ? `${label} must be greater than the minimum.` : null
}

// ── Files ───────────────────────────────────────────────────────────────────
export const fileLimits = (file, { maxMB, exts, label = 'File' }) => {
  if (!file) return `${label} is required.`
  const mb = file.size / (1024 * 1024)
  if (file.size === 0) return `${label} is empty.`
  if (maxMB && mb > maxMB) return `${label} is ${mb.toFixed(1)} MB — must be ≤ ${maxMB} MB.`
  if (exts) {
    const ext = _ext(file.name)
    if (!exts.includes(ext)) {
      return `${label}: extension "${ext || '?'}" is not allowed. Allowed: ${exts.join(', ')}`
    }
  }
  return null
}

export const imageFile = (file, { maxMB = FILE.LIMITS_MB.image, label = 'Image' } = {}) =>
  fileLimits(file, { maxMB, exts: FILE.IMAGE_EXTS, label })

export const qrFile = (file) =>
  fileLimits(file, { maxMB: FILE.LIMITS_MB.qr, exts: FILE.IMAGE_EXTS, label: 'QR image' })

export const slipFile = (file) =>
  fileLimits(file, {
    maxMB: FILE.LIMITS_MB.slip,
    exts: [...FILE.IMAGE_EXTS, ...FILE.PDF_EXTS],
    label: 'Slip',
  })

export const attachmentFile = (file) =>
  fileLimits(file, {
    maxMB: FILE.LIMITS_MB.attachment,
    exts: [...FILE.IMAGE_EXTS, ...FILE.PDF_EXTS, ...FILE.DOC_EXTS],
    label: 'Attachment',
  })

// ── Helpers ─────────────────────────────────────────────────────────────────
/**
 * Run a rule map against a values object and return an errors object.
 * rules: { fieldName: [validatorFn, ...] }
 * Returns: { fieldName: 'first error', ... }  (only fields with errors)
 */
export function validate(values, rules) {
  const out = {}
  for (const [field, fns] of Object.entries(rules)) {
    for (const fn of fns) {
      const err = fn(values[field])
      if (err) { out[field] = err; break }
    }
  }
  return out
}

/**
 * Map the standard `{success, errors}` API envelope into a flat field map.
 * Accepts axios errors: returns { fieldName: 'message', non_field: 'message' }.
 */
export function extractApiErrors(axiosErr) {
  const out = {}
  // No error at all (e.g. mutation hasn't been triggered yet) → empty map
  if (!axiosErr || (typeof axiosErr === 'object' && Object.keys(axiosErr).length === 0)) {
    return out
  }
  const data = axiosErr?.response?.data
  if (!data) {
    // Genuine network/transport failure
    out.non_field = axiosErr?.message || 'Network error. Please try again.'
    return out
  }
  const errs = data.errors
  if (errs && typeof errs === 'object') {
    for (const [k, v] of Object.entries(errs)) {
      out[k === 'non_field_errors' ? 'non_field' : k] = Array.isArray(v) ? v[0] : String(v)
    }
  }
  if (!Object.keys(out).length && data.message) {
    out.non_field = data.message
  }
  return out
}
